File: examples/python/control-flow/control_flow.py
# 第 11 章 · 流程控制 — Python 示例
# 运行：python3 control_flow.py

scores = [92, 75, 50]

# 1. 分支：用缩进表示层级，关键字是 elif
def grade(score):
    if score >= 90:
        return "A"
    elif score >= 60:
        return "B"
    else:
        return "C"
print("分支:", " ".join(grade(s) for s in scores))

# 2. Python 只有 for-each；需要下标用 enumerate，需要计数用 range
print("直接遍历:", [s for s in scores])
print("带下标:  ", [(i, s) for i, s in enumerate(scores)])
print("计数:    ", list(range(3)))

# 3. Python 独有的 for-else：循环正常跑完（没 break）才执行 else
def find(target, items):
    for x in items:
        if x == target:
            return f"找到 {x}"
    else:
        return "循环跑完也没找到（else 分支执行了）"
print("for-else:", find(75, scores))
print("for-else:", find(99, scores))

# 4. 卫语句
def process(user):
    if user is None:
        return "无用户"
    if not user.get("active"):
        return "未激活"
    return "处理完成"
print("卫语句:", process(None), "|", process({"active": True}))

# 5. 多分支：3.10 前用 dict 分派代替 switch
#    （Python 3.10+ 可用 match-case，本示例为兼容旧版未使用）
handler = {"A": "优秀", "B": "及格"}
print("dict 分派:", handler.get("A", "不及格"), handler.get("C", "不及格"))

File: examples/javascript/control-flow/control-flow.js
// 第 11 章 · 流程控制 — JavaScript 示例
// 运行：node control-flow.js

const scores = [92, 75, 50];

// 1. 分支
function grade(score) {
  if (score >= 90) return "A";
  else if (score >= 60) return "B";
  else return "C";
}
console.log("分支:", scores.map(grade).join(" "));

// 2. switch 需要 break，否则穿透
function describe(g) {
  switch (g) {
    case "A": return "优秀";
    case "B": return "及格";
    default:  return "不及格";
  }
}
console.log("switch:", describe("A"), describe("C"));

// 3. for...in 取键，for...of 取值 —— 最易混淆的一对
const keys = [], values = [];
for (const k in scores) keys.push(k);
for (const v of scores) values.push(v);
console.log("for...in 得到键:", keys, "| for...of 得到值:", values);

// 4. 经典陷阱：闭包捕获循环变量
const fnsVar = [], fnsLet = [];
for (var i = 0; i < 3; i++) fnsVar.push(() => i);
for (let j = 0; j < 3; j++) fnsLet.push(() => j);
console.log("用 var →", fnsVar.map(f => f()), "← 全是 3！");
console.log("用 let →", fnsLet.map(f => f()), "← 正确");

// 5. 卫语句：消除嵌套
function process(user) {
  if (!user) return "无用户";
  if (!user.active) return "未激活";
  return "处理完成";
}
console.log("卫语句:", process(null), "|", process({ active: true }));
